src/v/kafka/server/connection_context.cc

```diff
@@ -353,169 +355,143 @@ connection_context::reserve_request_units(api_key key, size_t size) {
+     * that invokes this behavior in the server: that is, it won't
+     * send auth data (or any other requests) until handshake or the
+     * full auth-process completes, etc... but representing these
+     * nuances of the protocol _explicitly_ in the server makes its
+     * behavior easier to understand and avoids misbehaving clients
+     * creating server-side errors that will appear as a corrupted
+     * stream at best and at worst some odd behavior.
+     */
+
+    const auto correlation = rctx.header().correlation;
+    const sequence_id seq = _seq_idx;
+    _seq_idx = _seq_idx + sequence_id(1);
+    auto res = kafka::process_request(
+      std::move(rctx), _server.smp_group(), *sres);
+    /**
+     * first stage processed in a foreground.
+     */
+    std::exception_ptr dispatch_eptr;
+    try {
+        co_await std::move(res.dispatched);
+    } catch (...) {
+        dispatch_eptr = std::current_exception();
+    }
+
+    /*
+     * if the dispatch/first stage failed, then we need to
+     * need to consume the second stage since it might be
+     * an exceptional future. if we captured `f` in the
+     * lambda but didn't use `then_wrapped` then the
+     * lambda would be destroyed and an ignored
+     * exceptional future would be caught by seastar.
+     */
+    if (dispatch_eptr) {
+        try {
+            co_await std::move(res.response);
+        } catch (...) {
+            vlog(
+              klog.info,
+              "Discarding second stage failure {}",
+              std::current_exception());
+        }
+        sres->tracker->mark_errored();
+        vlog(
+          klog.info, "Detected error dispatching request: {}", dispatch_eptr);
+        self->conn->shutdown_input();
+        sres->tracker->mark_errored();
+        co_return;
+    }
+
+    /**
+     * second stage processed in background.
+     */
+    ssx::background = ssx::spawn_with_gate_then(_server.conn_gate(), [&] {
+        return handle_response(
+          self, std::move(res.response), sres, seq, correlation);
+    });
+}
+
+ss::future<> connection_context::handle_response(
+  ss::lw_shared_ptr<connection_context> self,
+  ss::future<response_ptr> f,
+  ss::lw_shared_ptr<session_resources> sres,
+  sequence_id seq,
+  correlation_id correlation) {
+    try {
+        auto r = co_await std::move(f);
+        r->set_correlation(correlation);
+        response_and_resources randr{std::move(r), sres};
+        _responses.insert({seq, std::move(randr)});
+        co_return co_await maybe_process_responses();
+    } catch (...) {
+        // ssx::spawn_with_gate already caught
+        // shutdown-like exceptions, so we should only
+        // be taking this path for real errors.  That
+        // also means that on shutdown we don't bother
+        // to call shutdown_input on the connection,
+        // so rely on any future reader to check the
+        // abort source before considering reading the
+        // connection.
```

Comment:
I don't think this is true anymore, given this is caught from within the spawn_with_gate_then

Reply:
good catch. i thought this comment didn't make sense before, but looking closer, i thought the original handle exception was inside the gate, but it was attached to the gate. so the comment is true, and i fix this PR to also eat those shutdown exceptions before we get to this point.

Reply:
fixed. got to learn about `auto r = co_await ssx::ignore_shutdown_exceptions(std::move(f));` which i didn't know existed.

Reply:
oops, i messed it up. will push again.

Reply:
fixed. nice find. even careful peeling off of the levels of continuations in these conversions is easy miss a level of indention :(